pkg/ingester/wal.go

```diff
@@ -0,0 +1,586 @@
+	// Map of user_id -> series_fingerprint -> *memorySeries
+	allSeries := make(map[string]map[uint64]*memorySeries)
+
+	// Build the series from the last checkpoint.
+	lastCheckpointDir, _, err := lastCheckpoint(dir)
+	if err != nil {
+		return allSeries, err
+	}
+
+	sr, err := wal.NewSegmentsReader(lastCheckpointDir)
+	if err != nil {
+		return allSeries, err
+	}
+
+	series, err := loadCheckpoint(wal.NewReader(sr))
+	if err != nil {
+		return allSeries, err
+	}
+
+	for _, s := range series {
+		chunkDesc, err := fromWireChunks(s.Chunks)
+		if err != nil {
+			return allSeries, err
+		}
+
+		if _, ok := allSeries[s.UserId]; !ok {
+			allSeries[s.UserId] = make(map[uint64]*memorySeries)
+		}
+
+		ms := newMemorySeries(pbLabelPairToLabels(s.Labels))
+		ms.setChunks(chunkDesc)
+		allSeries[s.UserId][s.Fingerprint] = ms
+	}
+
+	// Load the WAL and add the samples to the series loaded
+	// from the checkpoint.
+	sr, err = wal.NewSegmentsReader(dir)
+	if err != nil {
+		return allSeries, nil
+	}
+	records, err := loadWAL(wal.NewReader(sr))
+	if err != nil {
+		return allSeries, nil
+	}
+
+	samplePair := model.SamplePair{}
+	for _, r := range records {
+		seriesMap, ok := allSeries[r.UserId]
+		if !ok {
+			seriesMap = make(map[uint64]*memorySeries)
+			allSeries[r.UserId] = seriesMap
+		}
+
+		// Some series might not be present in the checkpoint,
+		// hence add them here.
+		for _, lbl := range r.Labels {
+			if _, ok := allSeries[r.UserId][lbl.Fingerprint]; !ok {
+				allSeries[r.UserId][lbl.Fingerprint] = newMemorySeries(pbLabelPairToLabels(lbl.Labels))
+			}
+		}
+
+		for _, sample := range r.Samples {
+			samplePair.Timestamp = model.Time(sample.Timestamp)
+			samplePair.Value = model.SampleValue(sample.Value)
+			if err := seriesMap[sample.Fingerprint].add(samplePair); err != nil {
+				// We can ignore memorySeriesError because duplicate (or) out-of-order samples are possible
+				// here because the WAL is not truncated to align with the checkpoint.
+				if _, ok := err.(*memorySeriesError); !ok {
+					return allSeries, err
+				}
+			}
+		}
+	}
+
+	return allSeries, nil
+}
+
+func loadCheckpoint(r *wal.Reader) (series []*Series, err error) {
+	for r.Next() {
+		rec := r.Record()
+		s := &Series{}
+		if err := proto.Unmarshal(rec, s); err != nil {
+			return nil, err
+		}
+		series = append(series, s)
+
+	}
+	return series, r.Err()
+}
+
+func loadWAL(r *wal.Reader) (records []*Record, err error) {
+	for r.Next() {
+		rec := r.Record()
+		record := &Record{}
+		if err := proto.Unmarshal(rec, record); err != nil {
+			return nil, err
+		}
+		records = append(records, record)
+
+	}
+	return records, r.Err()
+}
+
+func pbLabelPairToLabels(lps []client.LabelPair) labels.Labels {
+	lbls := make(labels.Labels, len(lps))
+	for i := range lps {
+		lbls[i].Name = string(lps[i].Name)
+		lbls[i].Value = string(lps[i].Value)
+	}
+	return lbls
+}
+
+func recoverFromWal(ctx context.Context, ingester *Ingester, walDir string) (err error) {
```

Comment:
Wal should be WAL

In this PR it looks like there is no way to recover in operation, this is only used in tests?

Reply:
https://github.com/cortexproject/cortex/pull/1103#issuecomment-521627113

Reply:
I have updated the PR with recovery-mode ingester as per the doc https://docs.google.com/document/d/1n1HcdgmsqaVwqxKg_POZqIREOSqZIAlrxuyx4YYV_pA/edit#heading=h.np21hypvn83